client/task_runner.go

```diff
@@ -1442,7 +1442,7 @@ func (r *TaskRunner) updateServices(d driver.Driver, h driver.ScriptExecutor, ol
 		// Allow set the script executor if the driver supports it
 		exec = h
 	}
-	interpolateServices(r.getTaskEnv(), r.task)
+	interpolateServices(r.getTaskEnv(), new)
```

Comment:
Won't `r.getTaskEnv()` be materializing the wrong environment variables as well.

The task is only updated after this call on the task_runner: https://github.com/hashicorp/nomad/pull/2591/files#diff-e14e934fe7956cfba40a0a0f1419a173L1434